dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-fee-record-payment-entity-groups-to-keying-sheet.ts

```diff
@@ -1,65 +1,88 @@
 import Big from 'big.js';
-import { FeeRecordEntity, FeeRecordStatus, KeyingSheetAdjustment } from '@ukef/dtfs2-common';
+import { orderBy } from 'lodash';
+import { FeeRecordEntity, FeeRecordStatus } from '@ukef/dtfs2-common';
 import { FeeRecordPaymentEntityGroup } from '../../../../../helpers';
-import { KeyingSheet, KeyingSheetRow } from '../../../../../types/fee-records';
-import { mapFeeRecordEntityToKeyingSheetRowStatus, mapFeeRecordEntityToReportedPayments } from '../../../../../mapping/fee-record-mapper';
+import { KeyingSheet, KeyingSheetFeePayment, KeyingSheetRow } from '../../../../../types/fee-records';
+import { mapFeeRecordEntityToReportedPayments } from '../../../../../mapping/fee-record-mapper';
+import { mapFeeRecordEntityToKeyingSheetRowWithoutFeePayments, mapPaymentEntityToKeyingSheetFeePayment } from '../../../../../mapping/keying-sheet-mapping';
 
-const getKeyingSheetAdjustmentForAmount = (amount: number | null): KeyingSheetAdjustment | null => {
-  if (amount === null) {
-    return null;
-  }
+const STATUSES_OF_FEE_RECORDS_TO_DISPLAY_ON_KEYING_SHEET: FeeRecordStatus[] = ['READY_TO_KEY', 'RECONCILED'];
 
-  const amountAbsoluteValue = new Big(amount).abs().toNumber();
-  if (amountAbsoluteValue === 0) {
-    return {
-      amount: 0,
-      change: 'NONE',
-    };
-  }
+/**
+ * Creates keying sheet rows for fee record payment groups
+ * with many fee records and many payments by greedily
+ * splitting payment amounts across the fee record fees
+ * sorted in descending order
+ * @param param - The fee record payment entity groups
+ * @param param.feeRecords - The fee record entities
+ * @param param.payments - The payment entities
+ * @returns The keying sheet
+ */
+const getKeyingSheetRowsWithPaymentsAcrossFeeRecords = ({ feeRecords, payments }: FeeRecordPaymentEntityGroup): KeyingSheetRow[] => {
+  const feeRecordsSortedDescending = orderBy(feeRecords, [(feeRecord) => feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency()], ['desc']);
+  let allFeePaymentsSortedAscending = orderBy(payments, ['amount'], ['asc']).map(mapPaymentEntityToKeyingSheetFeePayment);
 
-  return {
-    amount: amountAbsoluteValue,
-    change: amount > 0 ? 'INCREASE' : 'DECREASE',
-  };
-};
+  return feeRecordsSortedDescending.map((feeRecord) => {
+    const feePaymentsForRow: KeyingSheetFeePayment[] = [];
+    let remainingFeeRecordAmount = new Big(feeRecord.getFeesPaidToUkefForThePeriodInThePaymentCurrency());
+    while (remainingFeeRecordAmount.gt(0)) {
+      const largestFeePayment = allFeePaymentsSortedAscending.pop()!;
+      feePaymentsForRow.push(largestFeePayment);
```

Comment:
I'm not sure that I like that we add the full payment and then reassign the amount later on, would it be possible to figure out the amount here and then add it with the correct amount?

Reply:
Given that I need to take tolerances into account, I will probably need to calculate the actual amount here anyway like you say. I originally didn't because it bloats the code quite a lot of perform that calculation in here